### PR TITLE
Supported new sample - VirtualList inline editor

### PR DESCRIPTION
**Pull request by KartnDev**
image
image

### CODE REVIEW

File: src/main/java/io/jmix/uisamples/view/entity/food/FoodDetailView.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        initIconFields();
+    }
+
+    @Subscribe("titleField")
+    public void onNameFieldTypedValueChange(final SupportsTypedValue.TypedValueChangeEvent<TypedTextField<String>, String> event) {
+        foodTitle.setText(event.getValue());
+    }
+
+    @Subscribe("foodIconUpload")
+    public void onFoodIconUploadFileUploadSucceeded(final FileUploadSucceededEvent<FileUploadField> event) {
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/entity/food/FoodDetailView.java:43]: Date binding has left the chat...


Replace with: 
```xml
<fileUploadField id="foodIconUpload" dataContainer="foodDc" property="icon" uploadIcon="UPLOAD" uploadText=""/>
```

After that....
`ItemPropertyChangeEvent` would be much better:
```java
    @Subscribe(id = "foodDc", target = Target.DATA_CONTAINER)
    protected void onPropertyChangeListener(InstanceContainer.ItemPropertyChangeEvent<Food> event) {
        if ("icon".equals(event.getProperty())) {
            initIconFields();
        }
    }
```

File: src/main/java/io/jmix/uisamples/view/entity/food/FoodDetailView.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    private void initIconFields() {
+        if (StringUtils.isNotBlank(getEditedEntity().getTitle())) {
+            foodTitle.setText(getEditedEntity().getTitle());
+        }
+        byte[] iconArray = getEditedEntity().getIcon();
+        if (iconArray != null && iconArray.length > 0) {
+            try {
+                String iconFileName = "%s.png".formatted(getEditedEntity().getTitle());
+                foodIcon.setImageResource(new StreamResource(iconFileName, () -> new ByteArrayInputStream(getEditedEntity().getIcon())));
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/entity/food/FoodDetailView.java:57]: Exceeds the 120 character limit

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    @Supply(to = "foodList", subject = "renderer")
+    private Renderer<Food> foodListRenderer() {
+        return new ComponentRenderer<>(item -> {
+            HorizontalLayout rootCardLayout = new HorizontalLayout();
+            rootCardLayout.setMargin(true);
+            rootCardLayout.setSpacing(true);
+
+            VerticalLayout infoLayout = new VerticalLayout();
+            infoLayout.setMargin(false);
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:70]: Margin is `false` by default, why are these lines (3 times)?

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    public void onBeforeShow(final BeforeShowEvent event) {
+        foodDl.load();
+    }
+
+    @Supply(to = "foodList", subject = "renderer")
+    private Renderer<Food> foodListRenderer() {
+        return new ComponentRenderer<>(item -> {
+            HorizontalLayout rootCardLayout = new HorizontalLayout();
+            rootCardLayout.setMargin(true);
+            rootCardLayout.setSpacing(true);
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:67]: Spacing is `true` by default??
image

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    private CollectionLoader<Food> foodDl;
+
+    @Subscribe
+    public void onBeforeShow(final BeforeShowEvent event) {
+        foodDl.load();
+    }
+
+    @Supply(to = "foodList", subject = "renderer")
+    private Renderer<Food> foodListRenderer() {
+        return new ComponentRenderer<>(item -> {
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:64]: Spaghetti is not the best way to organize code...
In the description of the sample, you described three renderer elements: `Avatar`, `InfoLayout` and `ButtonsLayout`.
Why not decompose the code and move the generation code of these three elements into separate methods??

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            buttonsPanel.setSpacing(false);
+
+            Button detailButton = new Button(new Icon(VaadinIcon.PENCIL));
+            detailButton.setText(messages.getMessage("actions.Edit"));
+            detailButton.addClickListener(e -> dialogWindows.detail(this, Food.class)
+                    .withViewClass(FoodDetailView.class)
+                    .editEntity(item)
+                    .withParentDataContext(getViewData().getDataContext())
+                    .withAfterCloseListener(closeEvent -> {
+                        FoodDetailView detail = closeEvent.getSource().getView();
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:109]: This variable is not used anywhere except in the `if` block. Why is it declared outside?

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                        if (closeEvent.closedWith(StandardOutcome.SAVE)) {
+                            foodDc.replaceItem(detail.getEditedEntity());
+                        }
+                    })
+                    .build()
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:114]: Why??

image

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        try (InputStream imageStream = resources.getResourceAsStream(iconLocation)) {
+            food.setIcon(imageStream.readAllBytes());
+        } catch (IOException e) {
+            throw new RuntimeException("Cannot open a stream", e);
+        }
+        return food;
+    }
+
+    private List<Food> getInitialFoodList() {
+        List<Food> foodList = new ArrayList<Food>();
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:157]: image

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                messages.getMessage(getClass(), "sample.fries.description"),
+                3, "META-INF/resources/icons/components/virtuallist-food/fries.png"));
+
+        foodList.add(createFoodSample(messages.getMessage(getClass(), "sample.pizza.title"),
+                messages.getMessage(getClass(), "sample.pizza.description"),
+                15, "META-INF/resources/icons/components/virtuallist-food/pizza.png"));
+
+        foodList.add(createFoodSample(messages.getMessage(getClass(), "sample.sandwich.title"),
+                NOOP_DESCRIPTION,
+                2, "META-INF/resources/icons/components/virtuallist-food/sandwich.png"));
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:176]: `META-INF/resources/icons/components` has already been used several times, maybe move it to a separate static variable? `ICONS_PATH_ROOT`?

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    @Subscribe("addAction")
+    public void onAddAction(final ActionPerformedEvent event) {
+        dialogWindows.detail(this, Food.class)
+                .withViewClass(FoodDetailView.class)
+                .newEntity()
+                .withParentDataContext(getViewData().getDataContext())
+                .withAfterCloseListener(closeEvent -> {
+                    FoodDetailView detail = closeEvent.getSource().getView();
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:187]: The same

File: src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                    if (closeEvent.closedWith(StandardOutcome.SAVE)) {
+                        foodDc.replaceItem(detail.getEditedEntity());
+                    }
+                })
+                .build()
```
Comment (KremnevDmitry) [inline src/main/java/io/jmix/uisamples/view/flowui/components/virtuallist/inlineeditor/VirtualListInlineEditor.java:192]: The same

**Review by KremnevDmitry — CHANGES_REQUESTED** [review]

File: src/main/resources/io/jmix/uisamples/messages_en.properties [hunk 12]
```diff
@@ -174,11 +180,32 @@ io.jmix.uisamples.view.entity.customer/customerDetailView.title=Customer
 
 # =========== Sample views ===========
 # ----------- CustomComponentView -----------
-io.jmix.uisamples.view.flowui.components.fileuploadfield.fileuploadfield/fileUploadField.title=File upload field
-
```
[the remark this answers is not in the document]
Comment (KartnDev) [reply src/main/resources/io/jmix/uisamples/messages_en.properties:178]: I deleted this message coz i didn't found any key usage... We can undo this deletion just for commit history purity, but i think in other way we need to delete it (I'm pretty sure this line is not used anywhere)
image

**Pull request merged**